src/Npgsql/NpgsqlCommand.cs

```diff
@@ -195,6 +195,26 @@ public override string CommandText
         }
     }
 
+    string GetBatchFullCommandText()
+    {
+        Debug.Assert(IsWrappedByBatch);
+        if (InternalBatchCommands.Count == 0)
+            return string.Empty;
+        if (InternalBatchCommands.Count == 1)
+            return InternalBatchCommands[0].CommandText;
+        // TODO: Potentially cache on connector/command?
+        var sb = new StringBuilder();
```

Comment:
Well, it's definitely not a bad thing to precalculate capacity before allocating `StringBuilder`, though iterating over each batch command (in addition to the one we already have) does make things a bit slower (whether it's still going to be faster compared to just leaving `StringBuilder` as is depends on a lot of things and will require a proper benchmark).
I think, probably the best thing here is to just make a predefined capacity (1024?). But since we're also going to backport this to 8.0 (and probably 7.0, didn't really look that closely), it's better to leave it as simple as possible and instead make a separate pr to improve things for 9.0.

Reply:
I'm also assuming that once we cache the StringBuilder on the connector, capacity issues no longer matter as the builder will grow once and then just reuse the internal memory etc.